metals/src/main/scala/scala/meta/internal/metals/Ammonite.scala

```diff
@@ -0,0 +1,390 @@
+
+  private val cancelables = new MutableCancelable()
+  private val isCancelled = new AtomicBoolean(false)
+  def cancel(): Unit = {
+    if (isCancelled.compareAndSet(false, true)) {
+      val buildShutdown = buildServer0 match {
+        case Some(build) => build.shutdown()
+        case None => Future.successful(())
+      }
+      try cancelables.cancel()
+      catch {
+        case NonFatal(_) =>
+      }
+      try buildShutdown.asJava.get(100, TimeUnit.MILLISECONDS)
+      catch {
+        case _: TimeoutException =>
+      }
+    }
+  }
+
+  def loaded(path: AbsolutePath): Boolean =
+    path.isAmmoniteScript && {
+      val uri = path.toNIO.toUri.toASCIIString
+      lastImportedBuild0.targetUris.contains(uri)
+    }
+
+  private def connectToNewBuildServer(
+      build: BuildServerConnection
+  ): Future[BuildChange] = {
+    scribe.info(s"Connected to Ammonite Build server v${build.version}")
+    cancelables.add(build)
+    compilers().cancel()
+    buildServer0 = Some(build)
+    val importedBuild0 = // FIXME timed("imported ammonite build") {
+      MetalsLanguageServer.importedBuild(build)
+    for {
+      build0 <- statusBar().trackFuture(
+        "Importing Ammonite scripts",
+        importedBuild0
+      )
+      _ = {
+        lastImportedBuild0 = build0
+      }
+      _ <- indexWorkspace()
+      // _ = checkRunningBloopVersion(build.version)
+      _ <- Future.sequence[Unit, List](
+        compilations()
+          .cascadeCompileFiles(buffers.open.toSeq) ::
+          compilers().load(
+            buffers.open.toSeq.filter(!_.isAmmoniteScript)
+          ) ::
+          Nil
+      )
+    } yield BuildChange.Reconnected
+  }
+
+  private def disconnectOldBuildServer(): Future[Unit] = {
+    if (buildServer0.isDefined)
+      scribe.info("disconnected: ammonite build server")
+    buildServer0 match {
+      case None => Future.successful(())
+      case Some(value) =>
+        // remove from cancellable too?
+        buildServer0 = None
+        lastImportedBuild0 = ImportedBuild.empty
+        cancelables.remove(value)
+        diagnostics().resetAmmoniteScripts()
+        value.shutdown()
+    }
+  }
+
+  def maybeImport(path: AbsolutePath): Unit =
+    if (path.isAmmoniteScript && !loaded(path))
+      languageClient()
+        .showMessageRequest(Messages.ImportAmmoniteScript.params())
+        .asScala
+        .onComplete {
+          case Failure(e) => scribe.warn("Error requesting Ammonite import", e)
+          case Success(resp) =>
+            resp.getTitle match {
+              case Messages.ImportAmmoniteScript.doImport =>
+                start(Some(path)).asScala.onComplete {
+                  case Failure(e) =>
+                    scribe.warn(s"Error importing Ammonite script $path", e)
+                  case Success(_) =>
+                }
+              case _ =>
+            }
+        }
+
+  def start(): CompletableFuture[Object] =
+    start(None)
```

Comment:
Why not. It's just a make-binary-compatibility-easier-to-preserve habit.

Reply:
We don't estimate it will ever be important to have it binary compatible. If needed we will expose an API, but since these packages are internal there are no guarantees for anyone who plans to use them outside of Metals.

Reply:
Fixed.